.github/workflows/build.yaml

```diff
@@ -1,26 +1,648 @@
+      steps:
+        - name: Create target directories
+          run: mkdir dockerfiles && mkdir build && chmod ugo+rwx build
+
+        - name: Download Tarball
+          uses: actions/download-artifact@v3
+          with:
+            name: tarball
+            path: build
+
+        - name: Download Dockerfiles
+          uses: actions/download-artifact@v3
+          with:
+            name: dockerfiles
+            path: dockerfiles
+
+        - name: Uncompress tarball
+          run: tar xvJf build/wake_*.tar.xz -C build
+
+        - name: Build docker image
+          run: docker build -f dockerfiles/wasm -t wake-wasm .
+
+        - name: Build
+          run: docker run -u build --rm --mount type=bind,source=$PWD/build,target=/build --device /dev/fuse --cap-add SYS_ADMIN --security-opt apparmor:unconfined wake-wasm /bin/sh -c 'make -C wake-* vscode'
+
+        - name: Install
+          run: install -D -t release/vscode build/wake-*/extensions/vscode/wake-*.vsix
+
+        - name: Upload
+          uses: actions/upload-artifact@v3
+          with:
+            name: release_vscode
+            path: release/vscode
+
+    wasm:
+      runs-on: ubuntu-latest
+      needs: tarball
+      name: WASM Build
+      steps:
+        - name: Create target directories
+          run: mkdir dockerfiles && mkdir build && chmod ugo+rwx build
+
+        - name: Download Tarball
+          uses: actions/download-artifact@v3
+          with:
+            name: tarball
+            path: build
+
+        - name: Download Dockerfiles
+          uses: actions/download-artifact@v3
+          with:
+            name: dockerfiles
+            path: dockerfiles
+
+        - name: Uncompress tarball
+          run: tar xvJf build/wake_*.tar.xz -C build
+
+        - name: Build docker image
+          run: docker build -f dockerfiles/wasm -t wake-wasm .
+
+        - name: Build
+          run: docker run -u build --rm --mount type=bind,source=$PWD/build,target=/build --device /dev/fuse --cap-add SYS_ADMIN --security-opt apparmor:unconfined wake-wasm /bin/sh -c 'make -C wake-* wasm'
+
+    alpine:
+      runs-on: ubuntu-latest
+      needs: tarball
+      name: Alpine Build
+      steps:
+        - name: Create target directories
+          run: mkdir dockerfiles && mkdir -p build/tests && chmod ugo+rwx build
+
+        - name: Download Tarball
+          uses: actions/download-artifact@v3
+          with:
+            name: tarball
+            path: build
+
+        - name: Download Test Artifacts
+          uses: actions/download-artifact@v3
+          with:
+            name: tests
+            path: build/tests
+
+        - name: Download Dockerfiles
+          uses: actions/download-artifact@v3
+          with:
+            name: dockerfiles
+            path: dockerfiles
+
+        - name: Uncompress tarball
+          run: tar xvJf build/wake_*.tar.xz -C build
+
+        - name: Set tests as executable
+          run: chmod -R +x build/tests
```

Comment:
Just trying to think of how to DRY things up, and for this step, why do we need to do this after extracting the artifact? I don't know how GitHub's artifacts work under the hood, but normally archive file formats (e.g. tar) should preserve the executability bit.

Reply:
> I don't know how GitHub's artifacts work under the hood, but normally archive file formats (e.g. tar) should preserve the executability bit.

It gets lost in in the upload flow :(, its a common enough question that it has a top level note in the documentation. https://github.com/actions/upload-artifact#permission-loss

Reply:
I had suspected and was afraid you were going to say that :(

Reply:
I do want to DRY here if possible.

The best option is via a `matrix` which basically repeats a job but with different parameters. And its _almost_ here with the following params.

- docker image
 - build command
 - test command
 - artificats glob
 - install destination
 - upload destination
 
 We could certainly just do extra steps like adding the tests/debian dirs even when they aren't needed to cover the differences, but with 6 parameters its hard for me to decided if its worth it. I'm personally leaning towards yes

Reply:
If you're willing to spend time to DRY it up, I'd support that effort, since with the 7 or so different environments, I think it'd be worth it. But if it ends up being too difficult, I don't think it would be the worst thing to leave things as they are right now.

Reply:
This should be done now!

Reply:
> > I don't know how GitHub's artifacts work under the hood, but normally archive file formats (e.g. tar) should preserve the executability bit.
> 
> It gets lost in in the upload flow :(, its a common enough question that it has a top level note in the documentation. https://github.com/actions/upload-artifact#permission-loss

[A few sections down](https://github.com/actions/upload-artifact#maintaining-file-permissions-and-case-sensitive-files), it looks like that might only apply to the raw uploads, and making a tarball is actually given as a solution for maintaining the permissions.  I'm sure you've included this line as a reaction to trying and failing to run the tests without it, so there's probably something else going on which is *also* stripping those permissions before the files get packaged up, but that just underscores how complex this whole remote-execution-via-ostensibly-data-files system is.  (Very handy though, no matter its foibles!)

Reply:
yeah adding it was in response to a failed test. We can explore improving it later, but it seems like a small issue and I'd like to get this landed to minimize the time with the CI in a weird state.